include/deal.II/base/patterns_tools.h

```diff
@@ -0,0 +1,682 @@
+   * or the entry is set to a new value.
+   *
+   * By default, the pattern to use is obtained by calling the function
+   * PatternsTools::Convert<T>::to_pattern(), but a custom one can be used.
+   */
+  template <class ParameterType>
+  void add_parameter(const std::string           &entry,
+                     ParameterType               &parameter,
+                     ParameterHandler            &prm,
+                     const std::string           &documentation = std::string(),
+                     const Patterns::PatternBase &pattern = *Convert<ParameterType>::to_pattern());
+
+  /**
+   * @addtogroup Exceptions
+   * @{
+   */
+
+  /**
+   * Exception.
+   */
+  DeclException2 (ExcNoMatch,
+                  std::string, PatternBase &,
+                  << "The string " << arg1 << " does not match the pattern \""
+                  << arg2.description() << "\"");
+  //@}
+}
+
+
+// ---------------------- inline and template functions --------------------
+namespace PatternsTools
+{
+  namespace internal
+  {
+    /**
+     * Store information about the rank types of the given class.
+     *
+     * A class has Rank equal to the number of different separators
+     * that are required to uniquely identify its element(s) in a string.
+     *
+     * This class is used to detect wether the class T is compatible
+     * with a Patterns::List pattern or with a Patterns::Map pattern.
+     *
+     * Objects like Point<dim> or std::complex<double> are vector-likes, and
+     * have vector_rank 1. Elementary types, like `int`, `unsigned int`, `double`, etc.
+     * have vector_rank 0. `std::vector`, `std::list` and in general containers have rank
+     * equal to 1 + vector_rank of the contained type. Similarly for map types.
+     *
+     * A class with list_rank::value = 0 is either elementary or a
+     * map. A class with map_rank::value = 0 is either a List compatible
+     * class, or an elementary type.
+     *
+     * Elementary types are not compatible with Patterns::List, but
+     * non elementary types, like Point<dim>(), or std::complex<double>, are
+     * compatible with the List type. Adding more compatible types is a matter
+     * of adding a specialization of this struct for the given type.
+     *
+     * @author Luca Heltai, 2017
+     */
+    template<class T, class Enable=void>
+    struct RankInfo
+    {
+      static constexpr int list_rank = 0;
+      static constexpr int map_rank = 0;
+    };
+  }
+
+  // Arithmetic types
+  template<class T>
+  struct Convert<T, typename std::enable_if<std::is_arithmetic<T>::value>::type>
+  {
+
+    static std::unique_ptr<PatternBase> to_pattern()
+    {
+      if (std::is_integral<T>::value)
+        return std_cxx14::make_unique<Integer>(std::numeric_limits<T>::min(), std::numeric_limits<T>::max());
+      else if (std::is_floating_point<T>::value)
+        return std_cxx14::make_unique<Double>(-std::numeric_limits<T>::max(), std::numeric_limits<T>::max());
```

Comment:
I think it would be useful to omit the min/max values here so that we just default to what these classes want to use, without trying to duplicate their default constructor values.